### PR TITLE
IPQ4019 - Add support for TP-Link Deco M5

### PR DESCRIPTION
**Pull request by oklona**
Add support for TP-Link Deco M5 versions 1.0, 2.0 and 3.0. V2.0 has different LED controller than 1.0/3.0, and therefore has its own DTS file and board defiition.

### CODE REVIEW

File: tools/firmware-utils/src/tplink-safeloader.c [hunk 1]
```diff
@@ -0,0 +1,3871 @@
+// SPDX-License-Identifier: BSD-2-Clause
```
Comment (hauke) [inline tools/firmware-utils/src/tplink-safeloader.c:1]: Please send a patch with your changes to the openwrt mailing list https://openwrt.org/submitting-patches

Comment (jwmullally) [reply tools/firmware-utils/src/tplink-safeloader.c:1]: firmware-utils was moved into a separate git repo here: https://git.openwrt.org/project/firmware-utils.git
I found it was easiest to check that out, make the commit locally, then use `git send-email` as described in *Sending patches by git send-email* above.
You can search the mailing list for "firmware-utils" examples of the commit messages people use
E.g. http://lists.openwrt.org/pipermail/openwrt-devel/2022-April/038421.html

**Comment by jwmullally** [comment]
You'll need to squash all the commits down to one and give it a proper commit message. Check previous TP-Link "new device" commits for examples. E.g. https://github.com/openwrt/openwrt/commit/c12ef2f76098923fc08ef9d7aa3d080b6e72dfb1 https://github.com/openwrt/openwrt/commit/9c335accfe57f3049850b8fdb684c9e7e5feea82

**Comment by oklona** [comment]
Thank you. I have started this now. I am also reviewing whether we actually need the awkward partitioning scheme that Deco M5 use, where several partitions have names ending with "@0" or "@1". Since this is only visible in the tplink-safeloader and ultimately, the factory image, I am testing whether different naming may work just as well.

File: target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata [hunk 2]
```diff
@@ -126,6 +126,10 @@ case "$FIRMWARE" in
 		caldata_extract_mmc "0:ART" 0x1000 0x2f20
 		ath10k_patch_mac $(mmc_get_mac_binary ARTMTD 0x0)
 		;;
+	tplink,deco-m5)
+		caldata_extract "0:ART" 0x1000 0x2f20
+		ath10k_patch_mac $(macaddr_add $(mtd_get_mac_binary "OPAQUE" 0x8) 2)
```
Comment (robimarko) [inline target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:131]: All of these should be replaceable with NVMEM

File: target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh [hunk 3]
```diff
@@ -25,6 +25,11 @@ preinit_set_mac_address() {
 		mac_lan=$(get_mac_binary "/sys/bus/i2c/devices/0-0050/eeprom" 0x66)
 		[ -n "$mac_lan" ] && ip link set dev eth0 address "$mac_lan"
 		;;
+	tplink,deco-m5)
+		base_mac=$(mtd_get_mac_binary "OPAQUE" 0x8)
```
Comment (robimarko) [inline target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh:29]: NVMEM as well

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+
+	chosen {
+		bootargs-append = " root=/dev/mtdblock16";
+	};
+
+	aliases {
+		spi0 = "/soc/spi@78b5000";
+		spi1 = "/soc/spi@78b6000";
+		i2c0 = "/soc/i2c@78b7000";
+		i2c1 = "/soc/i2c@78b8000";
```
Comment (robimarko) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts:23]: What are these used for?

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/soc/qcom,tcsr.h>
+#include <dt-bindings/leds/common.h>
+
+ / {
+	model = "TP-Link Deco M5";
+	compatible = "tplink,deco-m5", "qcom,ipq4019";
+
+	chosen {
+		bootargs-append = " root=/dev/mtdblock16";
```
Comment (robimarko) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts:16]: Why is this needed?
OpenWrt should find the root based on name

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+
+		reset {
+			label = "reset";
+			gpios = <&tlmm 18 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+
+	soc {
+		interrupt-controller@b000000 {
```
Comment (robimarko) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts:41]: Drop this, its generic to the SoC and already defined in the DTSI.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+			compatible = "qcom,msm-qgic2";
+			interrupt-controller;
+			#interrupt-cells = < 0x03 >;
+			reg = < 0xb000000 0x1000 0xb002000 0x1000 >;
+			linux,phandle = < 0x01 >;
+			phandle = < 0x01 >;
+		};
+
+		pinctrl@1000000 {
```
Comment (robimarko) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts:50]: Please use the existing labels whenever they exist and not put it under soc node.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+				pinmux_cs {
+					pins = "gpio12";
+					function = "gpio";
+				};
+				pinconf {
+					pins = "gpio13", "gpio14", "gpio15";
+					drive-strength = <12>;
+					bias-disable;
+				};
+				pinconf_cs {
```
Comment (robimarko) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts:81]: Merge this with pinmux_cs please

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+			spi_0_pins: spi_0_pinmux {
+				pinmux {
+					pins = "gpio13", "gpio14", "gpio15";
+					function = "blsp_spi0";
+				};
+				pinmux_cs {
+					pins = "gpio12";
+					function = "gpio";
+				};
+				pinconf {
```
Comment (robimarko) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts:76]: Merge this with pinmux for the same pins please

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+					output-high;
+				};
+			};
+
+			i2c_0_pins: i2c_0_pinmux {
+				pinmux {
+					pins = "gpio20", "gpio21";
+					function = "blsp_i2c0";
+				};
+				pinconf {
```
Comment (robimarko) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts:94]: Same thing, merge pinmux and pinconf

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+					pins = "gpio20", "gpio21";
+					drive-strength = <16>;
+					bias-disable;
+				};
+			};
+
+			wifi_pins: wifi_pinmux {
+				pinmux_1 {
```
Comment (robimarko) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts:102]: Are these LNA or any other amplifier enable pins or?

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+			status = "okay";
+		};
+
+		spi_0: spi@78b5000 {
+			pinctrl-0 = <&spi_0_pins>;
+			pinctrl-names = "default";
+			status = "okay";
+			cs-gpios = <&tlmm 12 GPIO_ACTIVE_HIGH>;
+
+			m25p80@0 {
```
Comment (robimarko) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts:137]: flash@0

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+				#address-cells = <1>;
+				#size-cells = <0>;
+				reg = <0>;
+				compatible = "jedec,spi-nor";
+				spi-max-frequency = <24000000>;
```
Comment (robimarko) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts:142]: What's the SPI-NOR model used?
Usually, they can go way above 24MHz, and 50MHz is the SPI controller max

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+				};
+			};
+		};
+
+		i2c_0: i2c@78b7000 {
+			pinctrl-0 = <&i2c_0_pins>;
+			pinctrl-names = "default";
+			status = "okay";
+				lp5521@32 {
+				compatible = "national,lp5521";
```
Comment (robimarko) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts:232]: Lacking indentation

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+
+        ess-psgmii@98000 {
+			status = "okay";
+		};
+
+        mdio@90000 {
+			status = "okay";
+		};
+
+		tcsr@1949000 {
```
Comment (robimarko) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-v2.dts:347]: Broken indenation

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5.dts [hunk 16]
```diff
@@ -0,0 +1,361 @@
...
+
+ 
+#include "qcom-ipq4019.dtsi"
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/soc/qcom,tcsr.h>
+
+/ {
+	model = "TP-Link Deco M5";
+	compatible = "tplink,deco-m5", "qcom,ipq4019";
```
Comment (robimarko) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5.dts:13]: This looks really similar to v2 DTS, can you make a share DTSI to avoid duplication.
Other than that, same comments from v2 DTS apply here

File: target/linux/ipq40xx/image/generic.mk [hunk 17]
```diff
@@ -980,6 +981,35 @@ define Device/tel_x1pro
...
+        IMAGE/factory.bin := append-rootfs | tplink-safeloader factory
+        IMAGE/sysupgrade.bin := append-kernel | append-rootfs | pad-rootfs | append-metadata
+endef
+TARGET_DEVICES += tplink_deco-m5
+
+define Device/tplink_deco-m5-v2
+        $(call Device/FitImage)
+        DEVICE_VENDOR := TP-Link
+        DEVICE_MODEL := Deco-M5
+		DEVICE_VARIANT := v2
```
Comment (robimarko) [inline target/linux/ipq40xx/image/generic.mk:1002]: Broken indent

File: target/linux/ipq40xx/image/generic.mk [hunk 18]
```diff
@@ -980,6 +981,35 @@ define Device/tel_x1pro
...
+        DEVICE_PACKAGES := ipq-wifi-tplink_deco-m5
+        TPLINK_BOARD_ID := DECO-M5
+        IMAGE_SIZE := 16640k
+        IMAGES += factory.bin
+        IMAGE/factory.bin := append-rootfs | tplink-safeloader factory
+        IMAGE/sysupgrade.bin := append-kernel | append-rootfs | pad-rootfs | append-metadata
+endef
+TARGET_DEVICES += tplink_deco-m5
+
+define Device/tplink_deco-m5-v2
```
Comment (robimarko) [inline target/linux/ipq40xx/image/generic.mk:998]: Can you please make a common profile and reuse it for both models instead of duplicating

File: target/linux/ipq40xx/patches-5.10/901-arm-boot-add-dts-files.patch [hunk 19]
```diff
@@ -47,6 +47,8 @@ Signed-off-by: John Crispin <john@phrozen.org>
  	qcom-ipq4019-ap.dk04.1-c3.dtb \
  	qcom-ipq4019-ap.dk07.1-c1.dtb \
  	qcom-ipq4019-ap.dk07.1-c2.dtb \
++	qcom-ipq4019-deco-m5.dtb \
```
Comment (robimarko) [inline target/linux/ipq40xx/patches-5.10/901-arm-boot-add-dts-files.patch:50]: Does this apply?

**Comment by robimarko** [comment]
I have left some comments after a brief review, please squash the commits and sort the comments.

**Comment by oklona** [comment]
Thanks, as a complete noob, I am afraid I don’t really know how to do this. I did see a similar  comment on a previous attempt to get this device included, and have tried a few different ways to do this, but none of them worked. There was also a mention that this should be possible directly from the DTS, but similarly, I didn’t find a method that worked. -So, if anyone has any hints as to how to do this, I would gladly implement 😊

 

 

From: Robert Marko ***@***.***>
Reply to: openwrt/openwrt ***@***.***>
Date: Monday, 25 April 2022 at 13:27
To: openwrt/openwrt ***@***.***>
Cc: oklona ***@***.***>, Author ***@***.***>
Subject: Re: [openwrt/openwrt] IPQ4019 - Add support for TP-Link Deco M5 (PR #9705)

 

@robimarko commented on this pull request.

In target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
> @@ -126,6 +126,10 @@ case "$FIRMWARE" in
               caldata_extract_mmc "0:ART" 0x1000 0x2f20
               ath10k_patch_mac $(mmc_get_mac_binary ARTMTD 0x0)
               ;;
+       tplink,deco-m5)
+              caldata_extract "0:ART" 0x1000 0x2f20
+              ath10k_patch_mac $(macaddr_add $(mtd_get_mac_binary "OPAQUE" 0x8) 2)
All of these should be replaceable with NVMEM

**Comment by robimarko** [comment]
You can find lot of examples in the following commit:
https://github.com/openwrt/openwrt/commit/cfc13c44595db591092859fc6adc71f1d8159c50

As well as by looking at the IPQ40xx target

**Comment by PolynomialDivision** [comment]
Please also add the DSA conversion:
https://github.com/openwrt/openwrt/pull/4721

File: tools/firmware-utils/src/tplink-safeloader.c (hunk 1 above)
Comment (oklona) [reply tools/firmware-utils/src/tplink-safeloader.c:1]: Ok, I am working on this separately. It should hopefully be approved soon.

**Comment by RolandoMagico** [comment]
Hello @oklona,
do you still work on this PR?
BR
Roland

**Comment by oklona** [comment]
Hello, @RolandoMagico 
No, unfortunately I haven't had time for a couple of years. -So I guess this PR can be closed.
